lib/connection.js

```diff
@@ -669,6 +669,12 @@ Connection.prototype.openUri = function(uri, options, callback) {
     if (typeof callback === 'function') {
       callback(null, this);
     }
+
+    this.then = (resolve) => {
```

Comment:
```suggestion
    this.then = (resolve) => {
      delete this.then;
      delete this.catch;
      resolve(this);
      return this
    }
```
if you don't do this, if you try to call `Promise.resolve(conn)` or `await conn` your interpreter will go in an infinite loop of calling `resolve` recursively as it tries to flatten the promise, see https://github.com/Automattic/mongoose/issues/8810

Reply:
makes, sense, fixed!

Reply:
I guess if you really wanted to you could do
```js
this.then = (resolve) => {
  process.nextTick(() => {
    delete this.then;
    delete this.catch;
    resolve(this);
  });
  return this;
}
```